Fetch all repository revisions

mvdbeek opened the pull request:
Add `return_all argument` to get_installed_repository, so that all installed
revisions are returned from the database. This reduces the amount of code
required to iteratively ask for installed_changeset_revision, followed by
changeset_revision, followed by other [older, newer] changeset revisions
retrieved from a toolshed.

This speeds up the "Install to Galaxy" process (and some other things, probably) by ~ 5X for https://toolshed.g2.bx.psu.edu/view/drosofff/suite_metavisitor_1_2/e24919521ffb.

```
With PR:
galaxy.webapps.galaxy.controllers.admin_toolshed DEBUG 2016-04-06 15:41:33,492 Get deps start
galaxy.webapps.galaxy.controllers.admin_toolshed DEBUG 2016-04-06 15:41:56,171 Get deps done

---
Without:
galaxy.webapps.galaxy.controllers.admin_toolshed DEBUG 2016-04-06 15:46:37,582 Get deps start
galaxy.webapps.galaxy.controllers.admin_toolshed DEBUG 2016-04-06 15:48:35,558 Get deps done
```

mvdbeek committed:
Fetch all repository revisions

Add return_all argument to get_installed_repository, so that all installed
revisions are returned from the database. This reduces the amount of code
required to iteratively ask for installed_changeset_revision, followed by
changeset_revision, followed by other [older, newer] changeset revisions
retrieved from a toolshed.

martenson commented:
I was digging in that code yesterday with similar thoughts, nice idea @mvdbeek

nsoranzo commented on lib/tool_shed/util/shed_util_common.py:
```diff
@@ -1212,32 +1210,29 @@ def repository_was_previously_installed( app, tool_shed_url, repository_name, re
+                                                      tool_shed = tool_shed,
+                                                      name = repository_name,
+                                                      owner = repository_owner,
+                                                      return_all = True)
```
Spaces around `=` in lines 1214-1217 should be removed to fix flake8 build.

nsoranzo commented on lib/tool_shed/util/shed_util_common.py:
```diff
@@ -877,7 +873,7 @@ def get_tool_shed_from_clone_url( repository_clone_url ):
+def get_installed_repository( app, tool_shed, name, owner, changeset_revision=None, installed_changeset_revision=None, return_all=False ):
     """
     Return a tool shed repository database record defined by the combination of a toolshed, repository name,
     repository owner and either current or originally installed changeset_revision.
```
Function doc should be updated with the new parameter.

martenson commented:
@nsoranzo the PR is wip; you might want to wait with doc comments on the review version :)

mvdbeek committed:
Add get_previous_changeset_revisions function

to avoid querying the database twice in
get_repository_for_dependency_relationship.
Also describe return_all argument in get_installed_repository docstring.

mvdbeek commented:
Thanks @nsoranzo @martenson.
It would be nice to get a review, but please hold back on a potential merge.
I think this will need some testing "in the wild".

mvdbeek commented on lib/tool_shed/util/shed_util_common.py:
```diff
@@ -604,36 +604,36 @@ def get_repository_for_dependency_relationship( app, tool_shed, name, owner, cha
+            for previous_changeset in changeset_revisions:
+                if previous_changeset in changeset_repo_d:
+                    return changeset_repo_d[ previous_changeset ]
+            # The received changeset_revision is no longer installable, so get the next changeset_revision
```
There is something falling through the cracks here:
Assuming the received changeset_revision is 2, and the installed (and current) changeset revision is 3, and the next installed changeset revision is 4 (because next_installable_changeset_revision returns the latest changeset revision for packages), we do not find the installed repository at changeset revision 3, even though I believe the logic is aiming for this.

This could be fixed by first asking for the next_changeset_revision, and then for previous_chanegset revision (relative to next_changeset_revision).
But if we do this, we might as well ignore the changeset_revision completely, and just return whatever repository matches toolshed/name/owner at the highest ctx_revision from the database, instead of sending requests to the toolshed.
Does anyone see any problem with either of those approaches? [I would obviously prefer the last]
Ping @martenson @davebx @gregvonkuster @nsoranzo 
If we go with the last approach we’ll probably have another significant speedup and reduction of code.

places where get_repository_for_dependency_relationship is used:

lib/tool_shed/metadata/metadata_generator.py —> handle_repository_elem()  would give nice speedup, effective behaviour anyway
lib/tool_shed/galaxy_install/installed_repository_manager.py —> get_dependencies_for_repository(): shouldn’t make any difference, or slightly more robust to changes in recent dependencies ?!
lib/tool_shed/galaxy_install/tool_dependencies/env_manager.py -> get_env_shell_file_paths(): will find latest package. more reliable, I guess ? 
lib/tool_shed/galaxy_install/tool_dependencies/recipe/tag_handler.py -> handle_complex_repository_dependency_for_package () -> would fix the problem mentioned above
lib/tool_shed/util/shed_util_common.py ->get_repository_ids_requiring_prior_import_or_install() : this is the effective behaviour anyway, reduction of code

gregvonkuster replied:
@mvdbeek Thanks for taking a look at this.  Its been a while since I was involved with this particular code, so please take my input with a grain of salt or two.  ;)

I assume the scenario that results in this function receiving an older changeset_revision than the one that is currently installed is that a repository is being installed that has a dependency on the older changeset_revision.  If this is the case, the older changeset_revision should be installed unless the repository type is tool_dependency_definition.  Since this function is called for all repository types, I think the repository type should be checked so that your new approach is used only for repositories of type tool_dependency_definition.  If the type is unrestricted, I think the original approach should be used because that scenario requires installation of the older changeset_revision.

I like the approach you've taken with the optional param for returning all installed repositories, no matter the changeset_revision.

mvdbeek commented:
Closing this for the moment, I'm almost sure this will have unintended consequences if merged as is.
